File: Projects/rosalind/KMER.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

public class KMER {
	private static int str2int(String str) {
		return IntStream.range(0, 4).map(x -> (int) Math.pow(4, 3-x) * "ACGT".indexOf(str.charAt(x))).sum();
	}
	public static void main(String[] args) throws IOException {
		Scanner in = new Scanner(System.in);
		String read = "";
		final int[] counts = new int[256];
		in.next();
		while (in.hasNext()) {
			read += in.next();
			for (int i = 0; i < read.length()-3; i++)
				counts[str2int(read.substring(i, i+4))]++;
			read = read.substring(read.length()-3);
		}
		System.out.println(Arrays.stream(counts).mapToObj(String::valueOf).collect(Collectors.joining(" ")));
	}
}

File: Projects/rosalind/KMER.kt
package rosalind

import java.io.IOException
import java.util.*
import java.util.stream.Collectors
import java.util.stream.IntStream

object KMER {
    private fun str2int(str: String): Int {
        return IntStream.range(0, 4).map { x: Int -> Math.pow(4.0, 3 - x.toDouble()).toInt() * "ACGT".indexOf(str[x]) }.sum()
    }

    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val stdin = Scanner(System.`in`)
        var read = ""
        val counts = IntArray(256)
        stdin.next()
        while (stdin.hasNext()) {
            read += stdin.next()
            for (i in 0 until read.length - 3) counts[str2int(read.substring(i, i + 4))]++
            read = read.substring(read.length - 3)
        }
        println(Arrays.stream(counts).mapToObj { i: Int -> i.toString() }.collect(Collectors.joining(" ")))
        stdin.close()
    }
}